File: core/src/snapshot.rs
use crate::{Card, CardView, ColorVec, DevelopmentCards, Noble, MAX_PLAYERS};
use serde::{Deserialize, Serialize};
use smallvec::SmallVec;

/// A struct to represent the game snapshot.
#[derive(Debug, Serialize, Deserialize)]
pub struct GameSnapshot {
    /// Is the game in the last round.
    pub last_round: bool,
    /// The current round.
    pub current_round: usize,
    /// The current player.
    pub current_player: usize,

    /// The tokens available in the game.
    pub tokens: ColorVec,
    /// The card pool snapshot.
    pub card_pool: CardPoolSnapshot,
    /// The nobles available in the game.
    pub nobles: SmallVec<Noble, { MAX_PLAYERS + 1 }>,

    /// The players' snapshot.
    pub players: SmallVec<PlayerSnapshot, MAX_PLAYERS>,
}

/// A struct to represent the card pool snapshot.
#[derive(Debug, Serialize, Deserialize)]
pub struct CardPoolSnapshot {
    /// The remaining cards in the pool.
    pub remaining: [usize; 3],
    /// The revealed cards in the pool.
    pub revealed: [SmallVec<Card, 4>; 3],
}

/// A struct to represent the player snapshot.
#[derive(Debug, Serialize, Deserialize)]
pub struct PlayerSnapshot {
    /// The index of the player.
    pub idx: usize,
    /// The number of points the player has.
    pub points: u8,
    /// The tokens the player has.
    pub tokens: ColorVec,
    /// The development cards the player has.
    pub development_cards: DevelopmentCards,
    /// The reserved cards the player has.
    pub reserved_cards: SmallVec<CardView, 3>,
    /// The nobles the player has visited.
    pub nobles: SmallVec<Noble, { MAX_PLAYERS + 1 }>,
}

File: sdk/py/splendor_actor/types/snapshot.py
from enum import Enum
from typing import List, Dict, Any, Tuple, Union

from .card import Card, ReservedCardType, DevelopmentCards, CardPool
from .color import Color, ColorVec
from .noble import Noble
from .tier import Tier
from .player import Player


class GameSnapshot:
    """
    represent the game state
    """

    last_round: bool
    current_round: int
    current_player: int
    tokens: ColorVec
    card_pool: CardPool
    nobles: List[Noble]
    players: List[Player]

    def __init__(
        self,
        last_round: bool,
        current_round: int,
        current_player: int,
        tokens: List[int],
        card_pool: Dict,
        nobles: List[Dict],
        players: List[Dict],
    ):
        self.last_round = last_round
        self.current_round = current_round
        self.current_player = current_player
        self.tokens = ColorVec(tokens)
        self.card_pool = CardPool(**card_pool)
        self.nobles = [Noble(**noble) for noble in nobles]
        self.players = [Player(**player) for player in players]

    @staticmethod
    def from_json(json: Dict) -> "GameSnapshot":
        """
        Convert a json to a GameSnapshot
        """
        return GameSnapshot(**json)
